docs/RFCS/range_leases.md

```diff
@@ -45,53 +45,68 @@ its replicas holding range leases at once?
-circular dependencies. This table maps node IDs to an epoch counter
-and a lease expiration timestamp.
-
-The range lease is moved from a special raft command (which writes to
-a range-local, non-transactional range lease key) to a transactional
-range-local key (similar to the range descriptor). The range lease
-identifies the node that holds the lease and its epoch counter. It has
-a start timestamp but does not include an expiration time. The range
-lease record is always updated in a distributed transaction with the
-node lease record to ensure that the epoch counter is consistent and
-the start time is greater than the prior range lease holder’s node
-lease expiration (plus the maximum clock offset).
-
-Each node periodically performs a conditional put to its node lease to
-increase the expiration timestamp and ensure that the epoch has not
-changed. If the epoch does change, *all* of the range leases held by
-this node are revoked. A node *must not* propose any commands with a
-timestamp greater than the latest expiration timestamp it has written
-to the node lease table.
+circular dependencies. This table maps node IDs to an epoch counter,
+and an expiration timestamp.
+
+## Liveness table
+
+Column|     Description
+NodeID|     node identifier
+Epoch|      monotonically increasing liveness epoch
+Expiration| timestamp at which the liveness record expires
+
+Each node periodically heartbeats the liveness table, which is
+implemented as a conditional put which increases the expiration
+timestamp and ensures that the epoch has not changed. If the epoch
+does change, *all* of the range leases held by this node are
+revoked. A node can only execute commands (propose writes to Raft or
+serve reads) if it's the range leaseholder, the range lease epoch is
+equal to the node's liveness epoch, and the command timestamp is less
+than the node's liveness expiration minus the maximum clock offset.
+
+The node liveness table is gossiped by the range lease holder for the
+range which contains it. Gossip is used in order to minimize fanout
+and make distribution efficient. The best-effort nature of gossip is
+acceptable here because timely delivery of node liveness updates are
+not required for system correctness. Any node which fails to receive
+liveness updates will simply resort to a conditional put to increment
+a seemingly not-live node's liveness epoch. The conditional put will
+fail because the expected value is out of date and the correct liveness
+info is returned to the caller.
 
 A range lease is valid for as long as the node’s lease has the same
-epoch. If a node is down (and its node lease has expired), another
-node may revoke its lease(s) by incrementing the node lease
+epoch. If a node is down (and its node liveness has expired), another
+node may revoke its lease(s) by incrementing the node liveness
 epoch. Once this is done the old range lease is invalidated and a new
 node may claim the range lease.
```

Comment:
I think it's nice to be explicit here about the disjointed invariant. A range lease can move from node1 to node2 only after the node1's liveness record has expired, and node2 has a valid unexpired liveness epoch.